src/session/content/item_row.rs

```diff
@@ -94,16 +98,290 @@ impl ItemRow {
+                                    "<b>{}</b>",
+                                    match message.chat().type_() {
+                                        ChatType::Supergroup(data) if data.is_channel() => {
+                                            gettext("Channel photo removed")
+                                        }
+                                        _ => {
+                                            if message.is_outgoing() {
+                                                gettext("You removed group photo")
+                                            } else {
+                                                gettext!("{} removed group photo", sender_name)
+                                            }
+                                        }
+                                    }
+                                ));
+                            }
+                            MessageContent::MessageChatJoinByRequest => {
+                                self.get_or_create_event_row().set_label(&format!(
+                                    "<b>{}</b>",
+                                    if message.is_outgoing() {
+                                        gettext("You joined the group")
+                                    } else {
+                                        gettext!("{} joined the group", sender_name)
+                                    }
+                                ));
+                            }
+                            MessageContent::MessageChatJoinByLink => {
+                                self.get_or_create_event_row().set_label(&format!(
+                                    "<b>{}</b>",
+                                    if message.is_outgoing() {
+                                        gettext("You joined the group via invite link")
+                                    } else {
+                                        gettext!("{} joined the group via invite link", sender_name)
+                                    }
+                                ));
+                            }
+                            MessageContent::MessageChatAddMembers(data) => {
+                                let my_user_id = message.chat().session().me().id();
+                                self.get_or_create_event_row().set_label(&format!(
+                                    "<b>{}</b>",
+                                    if message.sender().as_user().map(User::id).as_ref()
+                                        == data.member_user_ids.get(0)
+                                    {
+                                        if message.is_outgoing() {
+                                            gettext("You joined the group")
+                                        } else {
+                                            gettext!("{} joined the group", sender_name)
+                                        }
+                                    } else {
```

Comment:
This has changed a lot since the last time I reviewed it, and I see a few issues with the new code. I will go through these issues in the following comments.

The first suggestion would be to put this match arm in a dedicated function, as it has become very large.

Reply:
Do you mean `match` inside MessagePinMessage?

Reply:
> MessagePinMessage

Yes, exactly. From line 194 to line 271

Reply:
Please check the latest commit, I'll squash those commit if you like the result

Reply:
LGTM. Except there is still the typo "separete".